spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaHttpClientResponseSubscriber.java

```diff
@@ -74,130 +75,187 @@ public void onSubscribe(Subscription s) {
+            }
         }
-        return null;
-    }
 
-    private Subscription ensureSubscribed() {
-        return ensureSubscribed(subscription);
+        // Notify the publisher.
+        bodyPublisher.relayOnComplete(cause);
     }
 
-    private static <T> T ensureSubscribed(@Nullable T s) {
-        if (s == null) {
+    Subscription upstreamSubscription() {
+        final Subscription subscription = this.subscription;
+        if (subscription == null) {
             throw new IllegalStateException("No subscriber has been subscribed.");
         }
-        return s;
+        return subscription;
     }
 
     ResponseBodyPublisher toResponseBodyPublisher() {
-        if (!future.isDone()) {
+        if (!headersFuture.isDone()) {
             throw new IllegalStateException("HTTP headers have not been consumed yet.");
         }
-        final ResponseBodyPublisher publisher = new ResponseBodyPublisher(this);
-        this.publisher = publisher;
-        return publisher;
+        return bodyPublisher;
     }
 
-    final class ResponseBodyPublisher implements Publisher<HttpObject>, Subscription {
+    static final class ResponseBodyPublisher implements Publisher<HttpObject>, Subscription {
 
-        @Nullable
-        private Subscriber<? super HttpObject> subscriber;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<ResponseBodyPublisher, Subscriber> subscriberUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(ResponseBodyPublisher.class, Subscriber.class,
+                                                       "subscriber");
+
+        private final ArmeriaHttpClientResponseSubscriber parent;
 
-        private final ArmeriaHttpClientResponseSubscriber upstreamSubscriber;
+        @Nullable
+        private volatile Subscriber<? super HttpObject> subscriber;
 
-        private ResponseBodyPublisher(ArmeriaHttpClientResponseSubscriber upstreamSubscriber) {
-            this.upstreamSubscriber = upstreamSubscriber;
+        ResponseBodyPublisher(ArmeriaHttpClientResponseSubscriber parent) {
+            this.parent = parent;
         }
 
         @Override
         public void subscribe(Subscriber<? super HttpObject> s) {
-            subscriber = s;
+            for (;;) {
+                final Subscriber<? super HttpObject> oldSubscriber = subscriber;
+                if (oldSubscriber == null) {
+                    // The first subscriber.
+                    if (subscriberUpdater.compareAndSet(this, null, s)) {
+                        break;
+                    }
+                } else {
+                    // The second subscriber.
+                    final Subscriber<HttpObject> newSubscriber = new CompositeSubscriber<>(oldSubscriber, s);
+                    if (subscriberUpdater.compareAndSet(this, oldSubscriber, newSubscriber)) {
+                        break;
+                    }
+                }
+            }
+
             s.onSubscribe(this);
```

Comment:
So the `subscription.request(n)` will be called by the second subscribers as well which might give more demand that the main subscriber requested. Is it okay?

Reply:
Not okay. :sweat_smile: Fixed by using `Mono` of `HttpResponse.completionFuture()` for the late subscribers.